# Abstract
Using Tensorflow's Youtube tutorial: [intro to Machine Learning](https://www.youtube.com/watch?v=KNAWp2S3w94&ab_channel=TensorFlow) I create several machine learning models in Tensorflow.

In [1]:
# Imports for first two TensorFlow examples
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
import numpy as np

2022-10-28 15:41:20.851623: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [2]:
# Import to check model accuracy – does this only work for the first model, a 1-dim sequential model?
#from sklearn.metrics import accuracy_score
#accuracy_score(test_images, predictions)

## Model 1. Predict y-output for a given x-input

1. Build a Keras model, the simplest ml model, with one density (layer) and one unit (neuron)
2. Input random data (with a formula of 2x-1) through numpy
3. Fit a model with 500 epochs(generations)

In [3]:
model = keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])
model.compile(optimizer='sgd', loss='mean_squared_error')

xs = np.array([-1.0, 0.0, 1.0, 2.0, 3.0, 4.0], dtype=float)
ys = np.array([-3.0, -1.0, 1.0, 3.0, 5.0, 7.0], dtype=float)

model.fit(xs, ys, epochs=500)

print(model.predict([10.0]))

Epoch 1/500


2022-10-28 15:41:28.789545: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


1/1 [==============================] - 0s 382ms/step - loss: 56.2720
Epoch 2/500
1/1 [==============================] - 0s 4ms/step - loss: 44.6896
Epoch 3/500
1/1 [==============================] - 0s 7ms/step - loss: 35.5686
Epoch 4/500
1/1 [==============================] - 0s 5ms/step - loss: 28.3842
Epoch 5/500
1/1 [==============================] - 0s 4ms/step - loss: 22.7237
Epoch 6/500
1/1 [==============================] - 0s 7ms/step - loss: 18.2622
Epoch 7/500
1/1 [==============================] - 0s 9ms/step - loss: 14.7442
Epoch 8/500
1/1 [==============================] - 0s 4ms/step - loss: 11.9687
Epoch 9/500
1/1 [==============================] - 0s 5ms/step - loss: 9.7775
Epoch 10/500
1/1 [==============================] - 0s 4ms/step - loss: 8.0461
Epoch 11/500
1/1 [==============================] - 0s 5ms/step - loss: 6.6767
Epoch 12/500
1/1 [==============================] - 0s 4ms/step - loss: 5.5922
Epoch 13/500
1/1 [==============================] - 0s 5ms/step

1/1 [==============================] - 0s 7ms/step - loss: 0.2555
Epoch 105/500
1/1 [==============================] - 0s 4ms/step - loss: 0.2503
Epoch 106/500
1/1 [==============================] - 0s 5ms/step - loss: 0.2451
Epoch 107/500
1/1 [==============================] - 0s 4ms/step - loss: 0.2401
Epoch 108/500
1/1 [==============================] - 0s 3ms/step - loss: 0.2352
Epoch 109/500
1/1 [==============================] - 0s 3ms/step - loss: 0.2303
Epoch 110/500
1/1 [==============================] - 0s 3ms/step - loss: 0.2256
Epoch 111/500
1/1 [==============================] - 0s 4ms/step - loss: 0.2210
Epoch 112/500
1/1 [==============================] - 0s 3ms/step - loss: 0.2164
Epoch 113/500
1/1 [==============================] - 0s 3ms/step - loss: 0.2120
Epoch 114/500
1/1 [==============================] - 0s 3ms/step - loss: 0.2076
Epoch 115/500
1/1 [==============================] - 0s 3ms/step - loss: 0.2034
Epoch 116/500
1/1 [==============================] - 0

1/1 [==============================] - 0s 3ms/step - loss: 0.0308
Epoch 207/500
1/1 [==============================] - 0s 4ms/step - loss: 0.0301
Epoch 208/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0295
Epoch 209/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0289
Epoch 210/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0283
Epoch 211/500
1/1 [==============================] - 0s 4ms/step - loss: 0.0277
Epoch 212/500
1/1 [==============================] - 0s 2ms/step - loss: 0.0272
Epoch 213/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0266
Epoch 214/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0261
Epoch 215/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0255
Epoch 216/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0250
Epoch 217/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0245
Epoch 218/500
1/1 [==============================] - 0

1/1 [==============================] - 0s 3ms/step - loss: 0.0037
Epoch 309/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0036
Epoch 310/500
1/1 [==============================] - 0s 4ms/step - loss: 0.0036
Epoch 311/500
1/1 [==============================] - 0s 2ms/step - loss: 0.0035
Epoch 312/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0034
Epoch 313/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0033
Epoch 314/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0033
Epoch 315/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0032
Epoch 316/500
1/1 [==============================] - 0s 2ms/step - loss: 0.0031
Epoch 317/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0031
Epoch 318/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0030
Epoch 319/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0029
Epoch 320/500
1/1 [==============================] - 0

1/1 [==============================] - 0s 3ms/step - loss: 4.6482e-04
Epoch 409/500
1/1 [==============================] - 0s 2ms/step - loss: 4.5527e-04
Epoch 410/500
1/1 [==============================] - 0s 3ms/step - loss: 4.4592e-04
Epoch 411/500
1/1 [==============================] - 0s 3ms/step - loss: 4.3676e-04
Epoch 412/500
1/1 [==============================] - 0s 3ms/step - loss: 4.2779e-04
Epoch 413/500
1/1 [==============================] - 0s 3ms/step - loss: 4.1900e-04
Epoch 414/500
1/1 [==============================] - 0s 3ms/step - loss: 4.1040e-04
Epoch 415/500
1/1 [==============================] - 0s 3ms/step - loss: 4.0197e-04
Epoch 416/500
1/1 [==============================] - 0s 3ms/step - loss: 3.9371e-04
Epoch 417/500
1/1 [==============================] - 0s 3ms/step - loss: 3.8562e-04
Epoch 418/500
1/1 [==============================] - 0s 3ms/step - loss: 3.7770e-04
Epoch 419/500
1/1 [==============================] - 0s 4ms/step - loss: 3.6994e-04
Epoch 

## Model 2. Image Recognition

In [4]:
fashion_mnist = keras.datasets.fashion_mnist # import fashion images with 10 sections (index 9, what we're using, is shoes)
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)), # 28x28 is the size of the input images
    keras.layers.Dense(128, activation=tf.nn.relu), # this middle layer has 128 functions to adjust the output of the next layer
    keras.layers.Dense(10, activation=tf.nn.softmax) # this layer is 10 units 'wide' representing the number of different items of clothing in the dataset
])

model.compile(optimizer=tf.keras.optimizers.Adam(),
             loss='sparse_categorical_crossentropy')

model.fit(train_images, train_labels, epochs=5) # train the model

# test_loss, test_acc = model.evaluate(test_images, test_labels) # evaluate the model against the data which wasn't used (test data)

predictions = model.predict(test_images)

Epoch 1/5
1875/1875 [==============================] - 2s 1ms/step - loss: 3.1050
Epoch 2/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.7388
Epoch 3/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.6380
Epoch 4/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.5931
Epoch 5/5
313/313 [==============================] - 0s 715us/step


In [5]:
# understanding what the data looks like so I can update the code from my 3-4yo tutorial
print('test labels:', test_labels,len(test_labels))
print('train labels:', train_labels,len(train_labels))
print('test images:', len(test_images[0]), len(test_images[0][0]))

test labels: [9 2 1 ... 8 1 5] 10000
train labels: [9 0 0 ... 3 0 5] 60000
test images: 28 28


In [6]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten (Flatten)           (None, 784)               0         
                                                                 
 dense_1 (Dense)             (None, 128)               100480    
                                                                 
 dense_2 (Dense)             (None, 10)                1290      
                                                                 
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________


### Accuracy

In [7]:
# visualize prediction data
predictions[0]

array([1.08184520e-17, 5.91917236e-23, 2.22722329e-25, 8.06697492e-15,
       1.11073924e-13, 2.21376970e-01, 9.06838238e-16, 7.77997598e-02,
       8.74280317e-07, 7.00822473e-01], dtype=float32)

In [16]:
def accuracy_check(predictions, test_labels):
    count=0
    for i in range(len(test_labels)): # loop through all test labels

        best_guess = 0.
        for c in range(len(predictions[i])): # save the best guess' label in each prediction array by using the highest value
            if predictions[i][c] > best_guess:
                best_guess = predictions[i][c]
                predicted_label = c

        if predicted_label != test_labels[i]: # check the accuracy of the model
            count+=1

        #print(f'Predictions guessed label {predicted_label} for the test image {test_labels[i]}')
    print(f'Final Score: {(1-(count/10000))*100}%\nThere were {count} incorrect matches out of {len(test_labels)}.')

In [17]:
accuracy_check(predictions, test_labels)

Final Score: 78.02%
There were 2198 incorrect matches out of 10000.


### Saving and reloading

In [10]:
# Save the model into a folder
model.save('tfmodel')

# Delete the model
del model

# Load the model back into jupyter notebook
model = keras.models.load_model('tfmodel')

INFO:tensorflow:Assets written to: tfmodel/assets


### Revising the model for accuracy: more epochs

In [11]:
model2 = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)), # 28x28 is the size of the input images
    keras.layers.Dense(128, activation=tf.nn.relu), # this middle layer has 128 functions to adjust the output of the next layer
    keras.layers.Dense(10, activation=tf.nn.softmax) # this layer is 10 units 'wide' representing the number of different items of clothing in the dataset
])

model2.compile(optimizer=tf.keras.optimizers.Adam(),
             loss='sparse_categorical_crossentropy')

model2.fit(train_images, train_labels, epochs=50) # train the model

predictions2 = model2.predict(test_images)

Epoch 1/50
1875/1875 [==============================] - 2s 1ms/step - loss: 3.6382
Epoch 2/50
1875/1875 [==============================] - 2s 1000us/step - loss: 0.6744
Epoch 3/50
1875/1875 [==============================] - 2s 1ms/step - loss: 0.6071
Epoch 4/50
1875/1875 [==============================] - 2s 1ms/step - loss: 0.5753
Epoch 5/50
1875/1875 [==============================] - 2s 1ms/step - loss: 0.5597
Epoch 6/50
1875/1875 [==============================] - 2s 1ms/step - loss: 0.5498
Epoch 7/50
1875/1875 [==============================] - 2s 1ms/step - loss: 0.5371
Epoch 8/50
1875/1875 [==============================] - 2s 1ms/step - loss: 0.5364
Epoch 9/50
1875/1875 [==============================] - 2s 1ms/step - loss: 0.5336
Epoch 10/50
1875/1875 [==============================] - 2s 1ms/step - loss: 0.5336
Epoch 11/50
1875/1875 [==============================] - 2s 1ms/step - loss: 0.5196
Epoch 12/50
1875/1875 [==============================] - 2s 999us/step - loss: 0.5

In [18]:
accuracy_check(predictions2, test_labels)

Final Score: 81.63%
There were 1837 incorrect matches out of 10000.


### Revising the model for accuracy: more layers

In [19]:
model3 = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)), # 28x28 is the size of the input images
    keras.layers.Dense(128, activation=tf.nn.relu), # this middle layer has 128 functions to adjust the output of the next layer
    keras.layers.Dense(128, activation=tf.nn.relu), # ADDED IN THIRD MODEL
    keras.layers.Dense(10, activation=tf.nn.softmax) # this layer is 10 units 'wide' representing the number of different items of clothing in the dataset
])

model3.compile(optimizer=tf.keras.optimizers.Adam(),
             loss='sparse_categorical_crossentropy')

model3.fit(train_images, train_labels, epochs=5) # train the model

predictions3 = model3.predict(test_images)

Epoch 1/5
1875/1875 [==============================] - 3s 1ms/step - loss: 1.7471
Epoch 2/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.6503
Epoch 3/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.6070
Epoch 4/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.5388
Epoch 5/5
313/313 [==============================] - 0s 796us/step


In [20]:
accuracy_check(predictions3, test_labels)

Final Score: 82.25%
There were 1775 incorrect matches out of 10000.


### Revising the model for accuracy: a different optimizer

In [21]:
model4 = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)), # 28x28 is the size of the input images
    keras.layers.Dense(128, activation=tf.nn.relu), # this middle layer has 128 functions to adjust the output of the next layer
    keras.layers.Dense(10, activation=tf.nn.softmax) # this layer is 10 units 'wide' representing the number of different items of clothing in the dataset
])

model4.compile(optimizer=tf.keras.optimizers.SGD(),
             loss='sparse_categorical_crossentropy')

model4.fit(train_images, train_labels, epochs=5) # train the model

predictions4 = model4.predict(test_images)

Epoch 1/5
1875/1875 [==============================] - 2s 1ms/step - loss: 152.7746
Epoch 2/5
1875/1875 [==============================] - 2s 1ms/step - loss: 2.3027
Epoch 3/5
1875/1875 [==============================] - 2s 1ms/step - loss: 2.3026
Epoch 4/5
1875/1875 [==============================] - 2s 1ms/step - loss: 2.3026
Epoch 5/5
313/313 [==============================] - 0s 803us/step


In [22]:
accuracy_check(predictions4, test_labels)

Final Score: 10.009999999999996%
There were 8999 incorrect matches out of 10000.


### Revising the model for accuracy: even more epochs + 2 layers

In [25]:
model5 = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)), # 28x28 is the size of the input images
    keras.layers.Dense(128, activation=tf.nn.relu), # this middle layer has 128 functions to adjust the output of the next layer
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax) # this layer is 10 units 'wide' representing the number of different items of clothing in the dataset
])

model5.compile(optimizer=tf.keras.optimizers.Adam(),
             loss='sparse_categorical_crossentropy')

model5.fit(train_images, train_labels, epochs=100) # train the model

predictions5 = model5.predict(test_images)

Epoch 1/100
1875/1875 [==============================] - 3s 1ms/step - loss: 1.4446
Epoch 2/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.5349
Epoch 3/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.4624
Epoch 4/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.4282
Epoch 5/100
1875/1875 [==============================] - 3s 1ms/step - loss: 0.4079
Epoch 6/100
1875/1875 [==============================] - 3s 1ms/step - loss: 0.3871
Epoch 7/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3663
Epoch 8/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3580
Epoch 9/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3521
Epoch 10/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3455
Epoch 11/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3406
Epoch 12/100
1875/1875 [==============================] - 2s 1ms/step - lo

1875/1875 [==============================] - 2s 1ms/step - loss: 0.2504
Epoch 98/100
1875/1875 [==============================] - 3s 1ms/step - loss: 0.2375
Epoch 99/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2338
Epoch 100/100
313/313 [==============================] - 0s 915us/step


In [28]:
accuracy_check(predictions5, test_labels)

Final Score: 85.91%
There were 1409 incorrect matches out of 10000.


### Conclusions
- Epoch value: 5 --> 50, model accuracy 75% --> 82%.
- Add another layer, model accuracy 75% --> 82%.
- Add 2 layers, Epoch value: 5 --> 50, model accuracy 75% --> 86%.

- Switch optimizers Adam --> SGD, model accuracy 75% --> 10%.

More epochs certainly help, as it gives the model more time to reach the end of its gradient minimum. More layers may also allow the model to develop a more complex algorithm to estimate each image more accurately, although more testing is needed to confirm this. Perhaps more layers would hinder the model in some cases.

Optimizers are an issue that I need to develop a better understanding of, as I do not know why the Adam optimizer functiions so much better for this task than the SGD optimizer. I did notice, however, that the SGD loss calculations were displaying NAN.